lib/chef/provisioning/aws_driver/driver.rb

```diff
@@ -101,22 +113,128 @@ def allocate_load_balancer(action_handler, lb_spec, lb_options, machine_specs)
+            zone = subnet.availability_zone.name
+            actual_zones_subnets[zone] ||= Set.new
+            actual_zones_subnets[zone] << subnet.id
+          end
+          actual_elb.availability_zones.each do |zone|
+            actual_zones_subnets[zone.name] ||= Set.new
+          end
+
+          desired_zones_subnets = {}
+          (availability_zones || []).each do |zone|
+            desired_zones_subnets[zone.downcase] ||= Set.new
+          end
+          if subnets.empty?
+            subnet_objects = []
+          else
+            subnet_objects = ec2.client.describe_subnets(:subnet_ids => subnets)[:subnet_set]
+          end
+          subnet_objects.each do |subnet|
+            zone = subnet[:availability_zone].downcase
+            desired_zones_subnets[zone] ||= Set.new
+            desired_zones_subnets[zone] << subnet[:subnet_id]
+          end
+
+          # Adding a subnet automatically adds the availability zone (because subnets live in a zone)
+          # So add all necessary subnets first, then availability zones
+          attach_subnets = []
+          enable_zones = []
+          desired_zones_subnets.each do |zone, subnets|
+            if actual_zones_subnets.keys.include?(zone)
+              # The zone already exists - do we need to add any subnets?
+              attach_subnets += (subnets - actual_zones_subnets[zone]).to_a
+            else
+              # The zone doesn't exist - do we need to add it as a zone or via its subnets?
+              if subnets.empty?
+                enable_zones << zone
+              else
+                attach_subnets += subnets.to_a
+              end
+            end
+          end
+
+          # Add all available
+          unless enable_zones.empty?
+            perform_action.call("  enable availability zones #{enable_zones.join(', ')}") do
+              actual_elb.availability_zones.enable(*enable_zones)
+            end
+          end
+          unless attach_subnets.empty?
+            perform_action.call("  attach subnets #{attach_subnets.join(', ')}") do
+              elb.client.attach_load_balancer_to_subnets(
+                load_balancer_name: actual_elb.name,
+                subnets: attach_subnets
+              )
+            end
+          end
+
+          # Only remove an availability zone if removing all subnets AND removing the availabilty zone
+          detach_subnets = []
+          disable_zones = []
+          actual_zones_subnets.each do |zone, subnets|
+            if desired_zones_subnets.keys.include?(zone)
+              # Don't remove the subnet if it would remove a desired zone
+              unless desired_zones_subnets[zone].empty?
+                detach_subnets += (subnets - desired_zones_subnets[zone]).to_a
+              end
+            else
+              disable_zones << zone unless availability_zones.empty?
+              detach_subnets += subnets.to_a
+            end
+          end
```

Comment:
Lines 213-223 of this show my removal logic - this is where I made some assumptions.

1. If the past state specified 1 or more subnets in availability zone A, and the new state only specifies availability zone A (IE, switching from managing subnets to managing zones) then I don't remove any of the subnets.  I do this because there is no way to determine what the default subnet is from a given availability zone.  `subnets: ['123', '456']` -> `availability_zones: ['a']` = no change if subnets `123` and `456` are in zone A.
2. Similarly, if they move from `availability_zones: ['a']` -> `subnets: ['123', '456']` where `123` and `456` live in Zone A, I only add the missing subnets.
  1. However, if `123` is the default subnet and they move `availability_zones: ['a']` -> `subnets: ['456']` I *will* remove subnet `123`.  I do not touch the availability zone in this instance.
3. If they move from managing availability zones to managing subnets, I will only send add/remove subnet calls.  EG,  `availability_zones: ['a']` -> `subnets: ['789']` will send a 'remove subnet 123' and 'add subnet 789' call to AWS.  It will not send any add/remove calls for availability zones.
  1. However!  If they move from
```ruby
availability_zones: ['a']
```
to
```ruby
availability_zones: ['b']
subnets: ['789']
```
I will send AWS 'remove availability zone a', 'remove subnet 123', 'add availability zone b', 'add subnet 789'

Let me know if you don't agree with this logic.